src/Indice.Services/ViberServiceApifon.cs

```diff
@@ -0,0 +1,134 @@
+            }
+            return phone.ToString("D");
+        }).ToArray();
+        if (recipients.Any(phoneNumber => phoneNumber.Any(numberChar => !char.IsNumber(numberChar)))) {
+            throw new ArgumentException("Invalid recipients. Recipients cannot contain letters.", nameof(recipients));
+        }
+        var senderId = sender?.Id ?? Options.Sender ?? Options.SenderName;
+        var payload = new ApifonIMRequest(senderId, recipients, body);
+        payload.IMChannels.Add(new IMChannel {
+            SenderId = senderId,
+            Text = body
+        });
+        var signature = payload.Sign(Options.ApiKey, HttpMethod.Post.ToString(), SERVICE_ENDPOINT);
+        var request = new HttpRequestMessage {
+            Content = new StringContent(payload.ToJson(), Encoding.UTF8, MediaTypeNames.Application.Json),
+            Method = HttpMethod.Post,
+            RequestUri = HttpClient.BaseAddress
+        };
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.Json));
+        request.Headers.Add("X-ApifonWS-Date", payload.RequestDate.ToString("r"));
+        request.Headers.Authorization = new AuthenticationHeaderValue("ApifonWS", $"{Options.Token}:{signature}");
+        try {
+            Logger.LogInformation("The full request sent to Apifon: {requestMessage}", JsonSerializer.Serialize(request, GetJsonSerializerOptions()));
+            Logger.LogInformation("The following payload was sent to Apifon: {requestPayload}", payload.ToJson());
+            httpResponse = await HttpClient.SendAsync(request);
+        } catch (Exception ex) {
+            Logger.LogError("Viber/SMS Delivery took too long");
+            throw new SmsServiceException("Viber/SMS Delivery took too long", ex);
+        }
+        var responseString = await httpResponse.Content.ReadAsStringAsync();
+        if (!httpResponse.IsSuccessStatusCode) {
+            Logger.LogInformation("Viber/SMS Delivery failed. {statusCode} : {responseString}", httpResponse.StatusCode, responseString);
+            throw new SmsServiceException($"Viber/SMS Delivery failed. {httpResponse.StatusCode} : {responseString}");
+        }
+        response = JsonSerializer.Deserialize<ApifonResponse>(responseString, GetJsonSerializerOptions());
+        if (response.HasError) {
+            Logger.LogInformation("Viber/SMS Delivery failed. {responseStatus}. ResponseId: {responseId}", response.Status.Description, response.Id);
+            throw new SmsServiceException($"Viber/SMS Delivery failed. {response.Status.Description} responseId {response.Id}");
+        } else {
+            Logger.LogInformation("Viber/SMS message successfully sent: {result}", response.Results.FirstOrDefault());
+        }
+        messageId = response.Id;
+        var messageIds = response.Results?.Values.SelectMany(x => x?.Select(y => y.Id))?.ToList();
```

Comment:
var messageIdsQuery = response.Results?.Values.SelectMany(x => x?.Select(y => y.Id));
if (messageIdsQuery?.Count() > 0) {
    messageId = string.Join(",", messageIdsQuery.ToList()); // Enumerating when list is needed
}

Reply:
This will trigger twice the enumeration its not faster. but slower. Count on IEnumerable is enumerating in order to count.